docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,247 @@
+    2. ObjectMeta.DeletionTimestamp is updated from nil to non-nil.
+  * Applies the finalizing logic to the object in the update event.
+  * Removes itself from ObjectMeta.Finalizers after the finalizing logic is completed.
+  * The API server deletes the object after the last finalizer removes itself from the ObjectMeta.Finalizers field.
+  * Because it's possible for the finalizing logic to be applied multiple times, the finalizing logic has to be idempotent.
+  * If a finalizer fails to act in a timely manner, users can manually remove the finalizer from ObjectMeat.Finalizers by a regular kubectl command.
+
+## Changes to existing components
+
+* API server:
+  * Deletion handler:
+    * If the `ObjectMeta.Finalizers` of the object being deleted is non-empty, then updates the DeletionTimestamp, but does not delete the object.
+    * If the `ObjectMeta.Finalizers` is empty and the [options.GracePeriod](../../pkg/registry/generic/etcd/etcd.go#L405) is zero, then deletes the object. If the options.GracePeriod is non-zero, then just updates the DeletionTimestamp.
+  * Update handler:
+    * If the update removes the last finalizer, and the DeletionTimestamp is non-nil, and the DeletionGracePeriodSeconds is zero, then deletes the object from the registry.
+    * If the update removes the last finalizer, and the DeletionTimestamp is non-nil, but the DeletionGracePeriodSeconds is non-zero, then just updates the object.
+
+## 2B. Enforcing orphaning policty with a finalizer
+
+## API changes
+
+```
+type ObjectMeta struct {
+	...
+	ParentReferences []ObjectReference
+    Orphaning bool
+}
+```
+
+**ObjectMeta.ParentReferences**: same as in [design #1](#api-changes).
+**ObjectMeta.Orphaning**: indicates whether to orphan the children when this object is deleted.
+
+## New components
+
+* Garbage Collector:
+  * Almost identical to the [Garbage Collector](#new-components) in design #1.
+
+* Orphaning finalizer:
+  * Could be part of the garbage collector.
+  * Watches for update events as described in [2A](#new-components-1).
+  * If `ObjectMeta.Orphaning` of the updated object is true, removes the object from the `ParentReferences` of its children.
+    * children can be found via the DAG kept by GC, or by relisting the child resource and checking the ParentReferences field.
+  * Removes the finalizer from the `ObjectMeta.Finalizers` of the parent object.
+
+## Expected race conditions
+
+* *Controller of the parent resource creates a child object after the orphaning finalizer has updated the ParentReferences of other children*: This can be solved by implementing Status.ObservedGeneration in all resources, so that before applying the orphaning logic to children, the orphaning finalizer checks Status.ObservedGeneration to ensure the controller of the parent resource has observed the deletion. Note that in design #1, the same [race condition](#expected-race-conditions) is solved by introducing the tombstone.
```

Comment:
@caesarxuchao @lavalamp 

If the pods have been running for some time (say, minutes), it would not be good to delete them when they should have been orphaned.

What kind of pathological scenario are we concerned about, where the RC/RS controller could read and write pods, but the GC controller could not? Perhaps we need to confirm that the informer is able to contact the apiserver, and that the cache is not arbitrarily stale?

We've encountered a number of challenges with writing clients, especially controllers, that have convinced me that we need to add features to the API to make it easier to do so. For instance, with eventual consistency, without a per-object sequence number such as `generation`, it's not possible to determine, in general, when an object received back from the API contains changes made by the client, assuming some fields may have been changed by other actors, without successfully performing a write using resourceVersion.